README.md
# Flooph

Flooph is a Ruby library designed to let you take code from (possibly-malicious) users and evaluate it safely.
Instead of evaluating arbitrary Ruby code (or JavaScript, or any other interpreter), it specifies a custom 'language',
with its own parser and evaluation.

Flooph provides four core pieces of functionality:

* A simple syntax for specifying and saving key/value pairs (much like a Ruby Hash literal):

   ```ruby
   f = Flooph.new
   f.update_variables <<-USERINPUT
     name: "World"
     cats: 17
     dogs: cats + 1
     alive: yes
     trollLocation: "cave"
   USERINPUT
   ```

* A simple template language that supports conditional content and injecting content.

   ```ruby
   puts f.transform <<-ENDTEMPLATE
     Hello, {=name}!

     {?trollLocation="cave"}
     There is a troll glaring at you.
     {|}
     The air smells bad here, like rotting meat.
     {.}

     {?debug}Troll is at {=trollLocation}.{.}

     {?dogs>0}
     I own {=dogs} dogg{?dogs=1}y{|}ies{.} now.
     {.}

     {? cats=42 }
     I have exactly 42 cats! I'll never get more.
     {| cats=1 }
     I have a cat. If I get another, I'll have two.
     {| cats>1 }
     I have {=cats} cats.
     If I get another, I'll have {=cats+1}.
     {|}
     I don't have any cats.
     {.}
   ENDTEMPLATE
   #=> Hello, World!
   #=>
   #=> There is a troll glaring at you.
   #=>
   #=> I own 18 doggies now.
   #=>
   #=> I have 17 cats.
   #=> If I get another, I'll have 18.
   ```


* Standalone functionality for evaluating conditional expressions based on the key/values (also used in the templates).

   ```ruby
   f = Flooph.new
   f.update_variables <<-USERINPUT
    cats: 17
    dogs: cats + 1
    alive: yes
    trollLocation: "cave"
   USERINPUT
   f.conditional 'cats?'                     #=> true
   f.conditional 'cats > 0'                  #=> true
   f.conditional 'cats ≥ 42'                 #=> false
   f.conditional 'dogs = cats'               #=> false
   f.conditional 'trollLocation = "cave"'    #=> true
   f.conditional 'alive & (cats>0 | dogs>0)' #=> true
   ```

* Standalone functionality for evaluating value expressions based on the key/values (also used in the templates).

   ```ruby
   f = Flooph.new cats:17, dogs:25
   f.calculate 'cats + dogs'                 #=> 42
   ```

# Syntax Reference

## Template Syntax

**`{=expression}`** — value output

Calculate a value and place its result at this spot. You can have spaces inside and around the expression if you like, but the opening `{=` must not have a space between the characters.

The tag and expression must all be on the same line.


**`{?condition1}` contents1 `{.}`**  
**`{?condition1}` contents1 `{|}` else_contents `{.}`**  
**`{?condition1}` contents1 `{|condition2}` contents2 `{|condition3}` contents3 `{.}`**  
**`{?condition1}` contents1 `{|condition2}` contents2 `{|condition3}` contents3 `{|}` else_contents `{.}`**

Conditional content must start with a `{?…}` condition tag. _(See [Condition Syntax](#condition-syntax) below for details on writing a condition.)_ You can have spaces inside and around the condition if you like, but the opening `{?` must not have a space between the characters.

Conditional content must end with a `{.}` tag. This must not have spaces within it.

Between the `{?…}` and `{.}` you can have any number of `{|…}` additional conditions. The first condition that evaluates to true will show the content following that.

After any `{|…}` additional conditions you can end with a final `{|}` else condition. The content following this tag will be shown if no other conditions evaluate to true.

While the contents of each `{?…}` and `{|…}` tag must be on the same line, you can have newlines in the contents between tags.

Contents may be raw text, or may contain additional tags. For example:

    {?dogs>0}I own {=dogs} dogg{?dogs=1}y{|}ies{.} now.{.}


## Condition Syntax

Conditions use:

* Numeric and string comparisons, using `<  >  =  ==  ≤  <=  ≥  >=  ≠  !=`

   For example: `cat>0`, `loc="house"` (or `loc=="house"`), `dogs≠1` (or `dogs!=1`)

   Comparisons involving variables that do not have a value (`zxyp < 8`)
   or comparisons between invalid types (`cats>"!@$#"`) evaluate to `false`.

 * Variable presence/truthiness using just name (`alive`) or with optional trailing question mark (`alive?`).

 * Boolean composition using `|` (alternatively `||`) and `&` (alternatively `&&`), with parentheses to alter precedence, and leading `!` to negate the value.

    For example: `alive | lifeCount>3 & (deaths<5 | !undead) || !(f && g)`

   * `!foo` means "not foo", inverting the meaning
   * `&` has higher precedence than `|` (e.g. `a | b & c` is the same as `a | (b & c)`)

## Value Expressions
_TODO_


## Variable Assignment

_Example syntax_:

```yaml
pagesRead       : 0
wearingGlasses  : yes
where_is_the_cat: "outside"
```

Variables are set and updated with one variable per line, using the syntax…  
`variableName : valueExpression`  
…with as many spaces/tabs around the colon as you wish (or none).

Variable names must start with a latin letter (A-Z), and may then have any number of latin letters, numbers, or underscores. (Hyphens, spaces, and other characters are not allowed in a variable name.) Variable names are case-sensitive.

**Examples of Valid Variable Names**:

* `foo`
* `Foo` _(a different variable from `foo` or `FOO`)_
* `foo2`
* `the_GrEaTEst_variable___ever___`

**Examples of Invalid Variable Names**:

* `21guns` _(cannot start with a number)_
* `_secret_` _(cannot start with an underscore)_
* `so-cool` _(cannot contain a hypen)_
* `two words` _(cannot contain a space)_

Value expressions may reference other variables (including the same one). Expressions are evaluated from top to bottom; referencing a variable set later in the same assignment block will use its previous value (if any), while referencing a variable earlier will use the newly-set value.

For example, imagine that you initially set these variables:

```yaml
time: 0
a: 0
b: 0
```

Then, you update the variables using this variable assignment:

```yaml
time: time+1
a: b+1
b: a+1
```

After evaluation, the variables will have the values `time=1`, `a=1`, `b=2`. When evaluating the expression `b+1`, the value of `b` is `0`. The resulting value of `1` is used to update `a`. Then, when evaluating the expression `a+1` the value of `a` has already been updated to `1`, and so `b` gets set to `2`.


# Catching Errors

When a parse fails, Flooph will raise a `Parslet::ParseFailed` error. Rescue this error to react to the parse failure in your own code. See [this Parslet documentation](http://kschiess.github.io/parslet/tricks.html#toc-error-reports) for information on how to get detailed information on what parslet was expecting when the parse failed, and what it saw instead.


# Known Limitations (aka TODO)
_In decreasing priority…_

- Tests are ad-hoc, and not integrated into any release process.
- Value expression language is a little _too_ simple; should support full arithmetic, including parens.
- Missing benchmarking to try speed improvements.
- Supplying custom `vars` to methods updates it for the instance; perhaps provide a one-off method.


# License & Contact
Flooph is copyright ©2018 by Gavin Kistner and is licensed under the [MIT License][1]. See the `LICENSE` file for more details.

For bugs or feature requests please open [issues on GitHub][2]. For other communication you can [email the author directly](mailto:gavin@phrogz.net?subject=Flooph).

[1]: http://opensource.org/licenses/MIT
[2]: https://github.com/Phrogz/Flooph/issues


CHANGES.md
# History of Changes to Flooph

## v0.2 — 2018-Apr-20

* Unless `$DEBUG` is set, Flooph will no longer rescue parse errors and output debug information. On a parse failure `Parslet::ParseFailed` is raised.

## v0.1.3 — 2018-Apr-14

* Workaround [parslet bug #193](https://github.com/kschiess/parslet/issues/193)

## v0.1.1 — 2018-Apr-13

* Allow CRLF for newlines in variable assignments (silly HTML textarea)

## v0.1 — 2018-Apr-8

* Initial Release

test.rb
require_relative 'lib/flooph'

def run!
  test_conditional
  test_assignments
  test_template
  test_interactions
  test_failures
end

def test_interactions
  f = Flooph.new  # No variables yet
  f.update_variables <<-END
    debug: false
    cats: 17
    alive: yes
    trollLocation: "cave"
  END
  f.conditional "cats > 3"  # true
  f.update_variables "cats: cats + 1"
  p f.calculate("cats")
end

def test_assignments
  tests = {
    '
    foo:true
    name: "Gavin"
    long: "Gavin Kistner"
    hasSeen: no
    bar: 17
    gor: kbo+kbo
    yay: jim + 0
    jim: jim + 1
    sam: 1 + bar
    jam : bar - 1
    jar : jam
    zoo : jam - 16
    cats: -1.3431
    diff: 10-8
    ' => {
      foo: true,
      name: 'Gavin',
      long: 'Gavin Kistner',
      hasSeen: false,
      bar: 17,
      jim: 2,
      yay: 1,
      sam: 18,
      jam: 16,
      jar: 16,
      cats: -1.3431,
      diff: 2,
      zoo: 0,
      gor: nil
    },
    "foo:1\r\nbar:2"     => { foo:1, bar:2, jim:1 },
    "foo:1\n\n\nbar:2"   => { foo:1, bar:2, jim:1 },
    "foo:1\r\n\r\nbar:2" => { foo:1, bar:2, jim:1 },
  }

  f = Flooph.new
  tests.each do |str, expected|
    f.vars = {}
    actual = f.update_variables(str, {jim:1} )
    compare_hashes expected, actual
  end
end

def test_conditional

  tests = {
    'a & b & c'         => true,
    '!a & b & c'        => false,
    'foo < 100'         => true,
    'foo < 100.1'       => true,
    'foo < 42'          => false,
    'foo≤42 & foo≥42 && foo==42 & !(foo≠42)' => true,
    'cats < 12'         => false,
    'cats > 12'         => false,
    'cats = 12'         => false,
    'a | b | c'         => true,
    '!a | b | c'        => true,
    '!a | !b | !c'      => false,
    'foo'               => true,
    'foo?'              => true,
    'foo<17 | bar?'     => true,
    'cat?'              => false,
    '!foo'              => false,
    '(foo)'             => true,
    '!(foo)'            => false,
    'a<b && b<c'        => true,
    'a>b && b<c'        => false,
    'a & no'            => false,
    'no & a'            => false,
    '(a & no)'          => false,
    ' (a & no) & (b|c)' => false,
    '!(a & no) & (b&c)' => true,
    '(a & no) | (b|c)'  => true,
    '(a & no) | (b&c)'  => true,
    'name="Phrogz"'     => true,
    'name="Gavin"'      => false,
    'a>"Phrogz"'        => false,
    'a="Phrogz"'        => false,
    'a<"Phrogz"'        => false,
    'name<"Phrogz"'     => false,
    'name>"Phrogz"'     => false,
    'name<"ZZZ"'        => true,
    'name<"AAA"'        => false,
    'name>"ZZZ"'        => false,
    'name>"AAA"'        => true,
    '"Phrogz">a'        => false,
    '"Phrogz"=a'        => false,
    '"Phrogz"<a'        => false,
    '"Phrogz"=name'     => true,
    '"Gavin"=name'      => false,
    '"Phrogz"<name'     => false,
    '"Phrogz">name'     => false,
    '"ZZZ"<name'        => false,
    '"AAA"<name'        => true,
    '"ZZZ">name'        => true,
    '"AAA">name'        => false,
  }

  f = Flooph.new
  tests.each do |expression, expected|
    f.vars = {name:'Phrogz', foo:42, bar:true, a:1, b:2, c:3}
    value = f.conditional(expression)
    unless value==expected
      puts expression
      puts "expected:#{expected} got:#{value}"
      puts
    end
  end
end

def compare_hashes(h1,h2)
  unless (missing = h1.keys - h2.keys).empty?
    puts "Missing keys: #{missing.join(', ')}"
  end
  unless (extra = h2.keys - h1.keys).empty?
    puts "Extra keys: #{extra.join(', ')}"
  end
  (h1.keys & h2.keys).each do |k|
    unless h1[k] == h2[k]
      puts "For #{k} expected #{h1[k].inspect} but got #{h2[k].inspect}"
    end
  end
end

def test_template
  tests = [
    { vars:{}, template:"a", expected:"a" },
    { vars:{a:17}, template:"a", expected:"a" },
    { vars:{a:17}, template:"{=a}", expected:"17" },
    { vars:{a:17}, template:"{?a}{=a}{.}", expected:"17" },
    { vars:{a:17}, template:"{? a }{=a}{.}", expected:"17" },
    { vars:{a:17}, template:"{?!a}{=a}{.}", expected:"" },
    { vars:{a:17}, template:"{?!a}{=a}{|}b{.}", expected:"b" },
    { vars:{a:1,b:2}, template:"{=a+b}", expected:"3" },
    { vars:{a:1,b:2}, template:"{=a + b}", expected:"3" },
    { vars:{a:1,b:2}, template:"{= a + b }", expected:"3" },
  ]

  f = Flooph.new
  tests.each do |h|
    f.vars = h[:vars]
    result = f.transform(h[:template])
    if result != h[:expected]
      puts "Variables: #{h[:vars].inspect}"
      puts "Template:  #{h[:template].inspect}"
      puts "Expected:  #{h[:expected].inspect}"
      puts "Result:    #{result.inspect}"
      puts
    end
  end
end

def test_failures
  f = Flooph.new a:1
  ["{=foo", "{=foo}{?bar", "{?foo}yay{?bar{.}", "{?foo}yay{?bar}{.}"].each do |str|
    begin
      f.transform str
      warn "Failed to raise error on invalid template #{str.inspect}"
    rescue Parslet::ParseFailed
    end
  end
end

# $t.vars = {cats:1, dogs:2}
# <<END.split("\n").each{ |s| p s=>$t.value(s) }
# 1 + 2
# cats + 1
# cats - 1
# cats + dogs
# 1 + dogs
# 1 - dogs
# END

run!

lib/flooph.rb
require 'parslet'

class Flooph < Parslet::Parser
  VERSION = '0.2'

  # The current values used when evaluating templates and conditionals.
  # Can also be updated by user input using #update_variables.
  attr_accessor :vars

  # Create a new, reusable template parser and evaluator.
  #
  # @param vars [Hash] symbol-to-values used in templates and conditional evaluations.
  def initialize(vars={})
    super()
    @vars = vars
  end

  # Evaluate a template like the following example, inserting content and
  # evaluating conditional branches. If you don't supply `vars` then the
  # existing values for the instance are used.
  #
  #     Hello, {=name}!                                # Insert values from variables.
  #                                                    #
  #     {?trollLocation="cave"}                        # Conditional based on boolean expressions.
  #     There is a troll glaring at you.               # See #conditional for how to write conditions.
  #     {|}                                            # Conditional 'else' clause.
  #     The air smells bad here, like rotting meat.    #
  #     {.}                                            # End of the if/else.
  #                                                    #
  #     {?debug}Troll is at {=trollLocation}.{.}       # Conditional based on if the variable exists (and isn't false)
  #                                                    #
  #     {?dogs>0}                                      #
  #     I own {=dogs} dogg{?dogs=1}y{|}ies{.} now.     # Conditionals can be inline.
  #     {.}                                            #
  #                                                    #
  #     {? cats=42 }                                   #
  #     I have exactly 42 cats! I'll never get more.   #
  #     {| cats=1 }                                    # Else-if for chained conditionals.
  #     I have a cat. If I get another, I'll have two. #
  #     {| cats>1 }                                    #
  #     I have {=cats} cats.                           #
  #     If I get another, I'll have {=cats+1}.         # Output can do simple addition/subtraction.
  #     {|}                                            #
  #     I don't have any cats.                         #
  #     {.}                                            #
  #
  # @param vars [Hash] variable values to use for this and future evaluations.
  #        If omitted, existing variable values will be used.
  #        (see also #vars and #update_variables)
  # @return [String] the template after transformation.
  def transform(str, vars=nil)
    parse_and_transform(:mkup, str, vars).tap do |result|
      result.gsub!(/\n{3,}/, "\n\n") if result
    end
  end

  # Evaluate simple conditional expressions to a boolean value, with variable lookup.
  # Examples:
  #
  #     cats?            # variable cats is set (and not set to `false` or `no`)
  #     cats = 42        # variable `cats` equals 42
  #     cats>0 & dogs>0  # if both variables are numbers greater than zero
  #
  # * Numeric and string comparisons, using any of `<  >  =  ==  ≤  <=  ≥  >=  ≠  !=`
  #   * Non-present variables or invalid comparisons always result in false
  # * Variable presence/truthiness using just name (isDead) or with optional trailing question mark (isDead?).
  # * Boolean composition, e.g. `a | b & c & (d | !e) || !(f && g)`
  #   * `!foo` means "not foo", inverting the meaning
  #   * `&` has higher precedence than `|`
  #   * `|` is the same as `||`; `&` is the same as `&&`
  #
  # @param vars [Hash] variable values to use for this and future evaluations.
  #        If omitted, existing variable values will be used.
  #        (see also #vars and #update_variables)
  # @return [true, false] the result of the evaluation.
  def conditional(str, vars=nil)
    parse_and_transform(:boolean_expression, str, vars)
  end

  # Parse a simple hash setup for setting and updating values. For example:
  #
  #     f = Flooph.new  # No variables yet
  #     f.update_variables <<-END
  #       debug: false
  #       cats: 17
  #       alive: yes
  #       trollLocation: "cave"
  #     END
  #     f.conditional "cats > 3"
  #     #=> true
  #     f.update_variables "oldCats:cats \n cats: cats + 1"
  #     f.calculate "cats"
  #     #=> 18
  #     f.calculate "oldCats"
  #     #=> 17
  #
  # Legal value types are:
  #
  # * Booleans:  `true`, `false`, `yes`, `no`
  # * Numbers:   `-3`, `12`, `3.1415`
  # * Strings:   `"foo"`, `"Old Barn"` _must use double quotes_
  # * Variables: `cats + 7` _only supports add/subtract, not multiplication_
  #
  # @param vars [Hash] initial variable values to base references on.
  #        If omitted, existing variable values will be used.
  # @return [Hash] the new variable values after updating.
  def update_variables(str, vars=nil)
    parse_and_transform(:varset, str, vars)
    @vars
  end

  # Evaluate an expression, looking up values and performing simple math.
  #
  #     f = Flooph.new cats:17, dogs:25
  #     p f.calculate("cats + dogs")
  #     #=> 42
  #
  # @param vars [Hash] variable values to use for this and future evaluations.
  #        If omitted, existing variable values will be used.
  #        (see also #vars and #update_variables)
  # @return [Hash] the new variable values after updating.
  def calculate(str, vars=nil)
    parse_and_transform(:value, str, vars)
  end

  # Common implementation for other methods
  # @!visibility private
  def parse_and_transform(root_rule, str, vars)
    @vars = vars if vars
    begin
      str = str.strip.gsub(/^[ \t]+|[ \t]+$/, '')
      tree = send(root_rule).parse(str)
      Transform.new.eval(tree, @vars)
    rescue Parslet::ParseFailed => error
      raise unless $DEBUG
      puts "Flooph failed to parse #{str.inspect}"
      puts error.parse_failure_cause.ascii_tree
      # TODO: catch transformation errors
    end
  end

  # template
  rule(:mkup) { (proz | spit.as(:proz) | cond).repeat.as(:result) }
  rule(:proz) { ((str('{=').absent? >> str('{?').absent? >> str('{|').absent? >> str('{.}').absent? >> any).repeat(1)).as(:proz) }
  rule(:spit) { str('{=') >> sp >> value >> sp >> str('}') }
  rule(:cond) do
    test.as(:test)  >> mkup.as(:out)                    >>
    (elif.as(:test) >> mkup.as(:out)).repeat.as(:elifs) >>
    (ells >> mkup.as(:proz)).repeat(0,1).as(:else)      >>
    stop
  end
  rule(:test) { str('{?') >> sp >> boolean_expression >> sp >> str('}') }
  rule(:elif) { str('{|') >> sp >> boolean_expression >> sp >> str('}') }
  rule(:ells) { str('{|}') }
  rule(:stop) { str('{.}') }

  # conditional
  rule(:boolean_expression) { orrs }
  rule(:orrs)  { ands.as(:and) >> (sp >> str('|').repeat(1,2) >> sp >> ands.as(:and)).repeat.as(:rest) }
  rule(:ands)  { bxpr.as(:orr) >> (sp >> str('&').repeat(1,2) >> sp >> bxpr.as(:orr)).repeat.as(:rest) }
  rule(:bxpr) do
                 ((var.as(:lookup) | num | text).as(:a) >> sp >> cmpOp.as(:cmpOp) >> sp >> (var.as(:lookup) | num | text).as(:b)) |
                 (str('!').maybe.as(:no) >> var.as(:lookup) >> str('?').maybe) |
                 (str('!').maybe.as(:no) >> str('(') >> sp >> orrs.as(:orrs) >> sp >> str(')'))
  end
  rule(:cmpOp) { (match['<>='] >> str('=').maybe) | match['≤≥≠'] | str('!=') }

  # assignment
  rule(:varset){ pair >> (str("\r").maybe >> str("\n") >> pair.maybe).repeat }
  rule(:pair)  { var.as(:set) >> sp >> str(':') >> sp >> value.as(:val) >> sp }
  rule(:value) { bool | adds | num | text | var.as(:lookup) }
  rule(:adds)  { (var.as(:lookup) | num).as(:a) >> sp >> match['+-'].as(:addOp) >> sp >> (var.as(:lookup) | num).as(:b) }
  rule(:sp)    { match[' \t'].repeat }

  # shared
  rule(:bool)  { (str('true') | str('false') | str('yes') | str('no')).as(:bool) }
  rule(:text)  { str('"') >> match["^\"\n"].repeat.as(:text) >> str('"') }
  rule(:num)   { (str('-').maybe >> match['\d'].repeat(1) >> (str('.') >> match['\d'].repeat(1)).maybe).as(:num) }
  rule(:var)   { match['a-zA-Z'] >> match('\w').repeat }
  rule(:sp)    { match[' \t'].repeat }

  # Workaround parslet bug that requires a root for inspection
  root :mkup

  class Transform < Parslet::Transform
    rule(proz:simple(:s)){ s.to_s }
    rule(test:simple(:test), out:simple(:out), elifs:subtree(:elifs), else:subtree(:elseout)) do
      if test
        out
      elsif valid = elifs.find{ |h| h[:test] }
        valid[:out]
      else
        elseout[0]
      end
    end
    rule(result:sequence(:a)){ a.join }

    # conditional
    rule(a:simple(:a), cmpOp:simple(:op), b:simple(:b)) do
      begin
        case op
          when '<'       then a<b
          when '>'       then a>b
          when '=', '==' then a==b
          when '≤', '<=' then a<=b
          when '≥', '>=' then a>=b
          when '≠', '!=' then a!=b
        end
      rescue NoMethodError, ArgumentError
        # nil can't compare with anyone, and we can't compare strings and numbers
        false
      end
    end
    rule(orr:simple(:value)) { value }
    rule(and:simple(:value)) { value }
    rule(orr:simple(:first), rest:sequence(:rest)) { [first, *rest].all? }
    rule(and:simple(:first), rest:sequence(:rest)) { [first, *rest].any? }
    rule(no:simple(:invert), orrs:simple(:val)) { invert ? !val : val }
    rule(no:simple(:invert), lookup:simple(:s)){ v = vars[s.to_sym]; invert ? !v : v }

    # assignment
    rule(set:simple(:var), val:simple(:val)){ vars[var.to_sym] = val }
    rule(a:simple(:a), addOp:simple(:op), b:simple(:b)) do
      x = a.is_a?(Parslet::Slice) ? vars[a.to_sym] : a
      y = b.is_a?(Parslet::Slice) ? vars[b.to_sym] : b
      if x.nil? || y.nil?
        nil
      else
        case op
          when '+' then x+y
          when '-' then x-y
        end
      end
    end

    # shared
    rule(lookup:simple(:s)) { vars[s.to_sym] }
    rule(num:simple(:str)) { f=str.to_f; i=str.to_i; f==i ? i : f }
    rule(bool:simple(:s)){ d = s.str.downcase; d=='true' || d=='yes' }
    rule(text:simple(:s)){ s.str }

    def eval(tree, vars)
      apply(tree, vars:vars)
    end
  end
end
